src/meta/src/hummock/manager/compaction.rs

```diff
@@ -102,4 +116,149 @@ impl HummockManager {
+
+                let mut generated_task_count = 0;
+                let mut existed_groups = groups.clone();
+                let mut meet_error = false;
+                let mut wait_compact_groups: HashSet<CompactionGroupId> = HashSet::default();
+
+                while generated_task_count < pull_task_count && !meet_error {
+                    let compact_ret = self
+                        .get_compact_tasks(std::mem::take(&mut existed_groups), selector)
+                        .await;
+
+                    match compact_ret {
+                        Ok((compact_tasks, trivial_tasks)) => {
+                            if compact_tasks.is_empty() {
+                                break;
+                            }
+                            generated_task_count += compact_tasks.len();
+                            for task in trivial_tasks {
+                                existed_groups.push(task.compaction_group_id);
+                                wait_compact_groups.insert(task.compaction_group_id);
+                            }
+                            for task in compact_tasks {
+                                let task_id = task.task_id;
+                                existed_groups.push(task.compaction_group_id);
+                                wait_compact_groups.insert(task.compaction_group_id);
+                                if let Err(e) =
+                                    compactor.send_event(ResponseEvent::CompactTask(task))
+                                {
+                                    tracing::warn!(
+                                        error = %e.as_report(),
+                                        "Failed to send task {} to {}",
+                                        task_id,
+                                        compactor.context_id(),
+                                    );
+                                    self.compactor_manager.remove_compactor(context_id);
+                                    meet_error = true;
+                                    break;
+                                }
+                            }
+                            existed_groups.sort();
+                            existed_groups.dedup();
+                        }
+                        Err(err) => {
+                            tracing::warn!(error = %err.as_report(), "Failed to get compaction task");
+                            break;
+                        }
+                    };
+                }
+                if generated_task_count < pull_task_count && !meet_error {
+                    // no compact_task to be picked
+                    for group in groups {
+                        self.compaction_state.unschedule(group, task_type);
+                    }
+                } else if !meet_error {
+                    for group in groups {
+                        if wait_compact_groups.contains(&group) {
```

Comment:
`wait_compact_groups` is used to store compaction group that can generate compact task in this round. Therefore, I think we should clear it at the beginning of the while loop at L139. Also, we need to add some documentation on what `wait_compact_groups` is used for. It is not stright-forward to understand for someone that is not familiar with the previous logic.

Reply:
But since we must make sure that not generate tasks more than `pull_task_count`, we can not clear it simply.